File: implementations/go/examples/internal/application/command/analyze_monthly_spending_handler.go
package command

import (
	"context"
	"fmt"
	"time"

	"github.com/example/account-service/internal/domain/account"
)

// AnalyzeMonthlySpendingCommand is the input SpendingAnalysisTaskController
// (interface/task/) builds by deserializing the Task Queue message payload.
// AnalysisMonth is in "2006-01" format, and
// SpendingAnalysisScheduler.EnqueueMonthlySpendingAnalysis sends "last
// month" as of enqueue time as-is in the payload — the same reason as
// ApplyDailyInterestCommand's Date: if the Consumer recomputed "which month"
// from the actual clock at processing time, a delayed/backlogged run could
// analyze the wrong month.
//
// Unlike nestjs's AnalyzeMonthlySpendingCommand (which carries all 4 boundary
// dates through the payload explicitly), only the identifying period string
// is carried here — spendingAnalysisPeriodRange below re-derives the 4
// boundaries purely from that fixed string, never from the processing-time
// clock, so the same safety property holds. This mirrors the existing Go
// idiom already used one file over: SendCardUsageStatementCommand carries
// only Period, and periodRange derives its [from, to) range from it.
type AnalyzeMonthlySpendingCommand struct {
	AnalysisMonth string
}

// spendingAnalysisBatchSize uses 500 for the same reason as
// interestBatchSize/cardStatementBatchSize.
const spendingAnalysisBatchSize = 500

// AnalyzeMonthlySpendingHandler is a system-triggered use case driven by the
// Task Queue — the ETL, in full: Extract (paginate every ACTIVE account,
// summarize its and the prior month's WITHDRAWAL transactions), Transform
// (account.NewSpendingAnalysis's %-change/trend calculation), Load (one row
// per account per month into spending_analysis). The output is a queryable
// read-model row, not a file — the value is precomputing an aggregate a
// client would otherwise have to re-derive from potentially many raw
// Transaction rows on every request.
type AnalyzeMonthlySpendingHandler struct {
	repo         account.Repository
	analysisRepo account.SpendingAnalysisRepository
}

func NewAnalyzeMonthlySpendingHandler(repo account.Repository, analysisRepo account.SpendingAnalysisRepository) *AnalyzeMonthlySpendingHandler {
	return &AnalyzeMonthlySpendingHandler{repo: repo, analysisRepo: analysisRepo}
}

// Handle saves and returns immediately per account — there is no Domain
// Event/Outbox involvement in this use case at all (SpendingAnalysis is a
// plain read-model row, not an Aggregate that raises events), so unlike
// ApplyDailyInterestHandler this has nothing to do with the Outbox path.
func (h *AnalyzeMonthlySpendingHandler) Handle(ctx context.Context, cmd AnalyzeMonthlySpendingCommand) error {
	monthStart, monthEnd, previousMonthStart, previousMonthEnd, err := spendingAnalysisPeriodRange(cmd.AnalysisMonth)
	if err != nil {
		return fmt.Errorf("analyze monthly spending: %w", account.ErrInvalidAnalysisMonth)
	}

	for page := 0; ; page++ {
		accounts, total, err := h.repo.FindAccounts(ctx, account.FindQuery{
			Status: []account.Status{account.StatusActive},
			Take:   spendingAnalysisBatchSize,
			Page:   page,
		})
		if err != nil {
			return fmt.Errorf("analyze monthly spending: find accounts: %w", err)
		}

		for _, a := range accounts {
			alreadyAnalyzed, err := h.analysisRepo.HasAnalysis(ctx, a.AccountID, cmd.AnalysisMonth)
			if err != nil {
				return fmt.Errorf("analyze monthly spending: has analysis: %w", err)
			}
			if alreadyAnalyzed {
				continue // Skip: this account was already analyzed for this month.
			}

			current, err := h.repo.SummarizeTransactions(ctx, account.SummarizeTransactionsQuery{
				AccountID:   a.AccountID,
				Type:        []account.TransactionType{account.TransactionTypeWithdrawal},
				CreatedFrom: monthStart,
				CreatedTo:   monthEnd,
			})
			if err != nil {
				return fmt.Errorf("analyze monthly spending: summarize current month: %w", err)
			}
			previous, err := h.repo.SummarizeTransactions(ctx, account.SummarizeTransactionsQuery{
				AccountID:   a.AccountID,
				Type:        []account.TransactionType{account.TransactionTypeWithdrawal},
				CreatedFrom: previousMonthStart,
				CreatedTo:   previousMonthEnd,
			})
			if err != nil {
				return fmt.Errorf("analyze monthly spending: summarize previous month: %w", err)
			}

			analysis := account.NewSpendingAnalysis(a.AccountID, cmd.AnalysisMonth, current.TotalAmount, current.Count, previous.TotalAmount)
			if err := h.analysisRepo.SaveAnalysis(ctx, analysis); err != nil {
				return fmt.Errorf("analyze monthly spending: save analysis: %w", err)
			}
		}

		if len(accounts) == 0 || (page+1)*spendingAnalysisBatchSize >= total {
			break
		}
	}
	return nil
}

// spendingAnalysisPeriodRange derives the [monthStart, monthEnd) and
// [previousMonthStart, previousMonthEnd) UTC calendar boundaries purely from
// analysisMonth ("2006-01") — mirrors nestjs's
// computePreviousSpendingAnalysisPeriod's calendar math (monthStart is the
// 1st of analysisMonth, monthEnd is the 1st of the next month,
// previousMonthStart is the 1st of the month before analysisMonth, and
// previousMonthEnd is monthStart), just re-derived from the fixed period
// string instead of "now" — the same trade-off periodRange (in
// send_card_usage_statement_handler.go) already makes.
func spendingAnalysisPeriodRange(analysisMonth string) (monthStart, monthEnd, previousMonthStart, previousMonthEnd time.Time, err error) {
	monthStart, err = time.Parse("2006-01", analysisMonth)
	if err != nil {
		return time.Time{}, time.Time{}, time.Time{}, time.Time{}, err
	}
	monthEnd = monthStart.AddDate(0, 1, 0)
	previousMonthStart = monthStart.AddDate(0, -1, 0)
	previousMonthEnd = monthStart
	return monthStart, monthEnd, previousMonthStart, previousMonthEnd, nil
}

File: implementations/fastapi/examples/src/account/application/command/analyze_monthly_spending_handler.py
from __future__ import annotations

from datetime import datetime

from ...domain.repository import AccountRepository
from ...domain.spending_analysis import SpendingAnalysis
from ...domain.spending_analysis_repository import SpendingAnalysisRepository

PAGE_SIZE = 200


def _add_months(year: int, month: int, delta: int) -> tuple[int, int]:
    total = year * 12 + (month - 1) + delta
    new_year, new_month = divmod(total, 12)
    return new_year, new_month + 1


def spending_analysis_period_range(analysis_month: str) -> tuple[datetime, datetime, datetime, datetime]:
    """Derives the [month_start, month_end) and [previous_month_start, previous_month_end)
    UTC calendar boundaries purely from analysis_month ("YYYY-MM") — mirrors nestjs's
    computePreviousSpendingAnalysisPeriod's calendar math, just re-derived from the fixed
    period string instead of "now", the same trade-off the go port's
    spendingAnalysisPeriodRange already makes. Since analysis_month itself was fixed at
    enqueue time (by the Scheduler), re-deriving its boundaries here — even if this Handler
    runs on a delayed/backlogged Task, possibly days later — always analyzes the same month.
    """
    year, month = (int(part) for part in analysis_month.split("-"))
    month_start = datetime(year, month, 1)
    next_year, next_month = _add_months(year, month, 1)
    month_end = datetime(next_year, next_month, 1)
    prev_year, prev_month = _add_months(year, month, -1)
    previous_month_start = datetime(prev_year, prev_month, 1)
    previous_month_end = month_start
    return month_start, month_end, previous_month_start, previous_month_end


class AnalyzeMonthlySpendingHandler:
    """The Command Service for the monthly spending-analysis ETL batch. Since this is a
    system-driven use case triggered once a month by the Task Queue
    (account.analyze-monthly-spending), there is no Command dataclass representing a user
    request (the same reason as ApplyDailyInterestHandler/SendMonthlyCardStatementHandler)
    — analysis_month is instead the one piece of state the Scheduler fixed at enqueue time.

    The ETL, in full: Extract (paginate every ACTIVE account, summarize its and the prior
    month's WITHDRAWAL transactions), Transform (SpendingAnalysis.create's %-change/trend
    calculation), Load (one row per account per month into spending_analysis). The output is
    a queryable read-model row, not a file — the value is precomputing an aggregate a client
    would otherwise have to re-derive from potentially many raw Transaction rows on every
    request.
    """

    def __init__(self, repo: AccountRepository, analysis_repo: SpendingAnalysisRepository) -> None:
        self._repo = repo
        self._analysis_repo = analysis_repo

    async def execute(self, analysis_month: str) -> int:
        month_start, month_end, previous_month_start, previous_month_end = spending_analysis_period_range(
            analysis_month
        )
        analyzed_count = 0
        page = 0

        while True:
            accounts, total = await self._repo.find_accounts(page=page, take=PAGE_SIZE, status=["ACTIVE"])
            if not accounts:
                break

            for account in accounts:
                already_analyzed = await self._analysis_repo.has_analysis(account.account_id, analysis_month)
                if already_analyzed:
                    continue

                current_count, current_total = await self._repo.summarize_transactions(
                    account.account_id, ["WITHDRAWAL"], month_start, month_end
                )
                _previous_count, previous_total = await self._repo.summarize_transactions(
                    account.account_id, ["WITHDRAWAL"], previous_month_start, previous_month_end
                )

                analysis = SpendingAnalysis.create(
                    account_id=account.account_id,
                    analysis_month=analysis_month,
                    total_amount=current_total,
                    transaction_count=current_count,
                    previous_total_amount=previous_total,
                )
                await self._analysis_repo.save_analysis(analysis)
                analyzed_count += 1

            if (page + 1) * PAGE_SIZE >= total:
                break
            page += 1

        return analyzed_count
